test/boost/json_test.cc

```diff
@@ -54,3 +68,163 @@ BOOST_AUTO_TEST_CASE(test_parsing_map_from_null) {
+    auto sz = expected.size();
+    sv = sv.substr(sz, sv.size() - sz);
+    return trim(sv);
+}
+
+void verify_parenthesis(std::string_view sv) {
+    static std::unordered_map<char, char> paren_map = {{'{', '}'}, {'[', ']'}, {'(', ')'}, {'<', '>'}};
+
+    char open = sv[0];
+    char close = sv[sv.size() - 1];
+    auto it = paren_map.find(open);
+    if (it == paren_map.end()) {
+        BOOST_FAIL(format("Unexpected delimiters: '{}' '{}'", open, close));
+    }
+    BOOST_REQUIRE_EQUAL(close, it->second);
+}
+
+template <std::ranges::range Range>
+void test_format_range(const char* desc, Range x, std::vector<std::string> expected_strings, std::string expected_delim = ",", bool expect_parenthesis = true) {
+    auto str = seastar::format("{}", x);
+    BOOST_TEST_MESSAGE(format("{}: {}", desc, str));
+
+    auto fmt_str = fmt::format("{}", x);
+    BOOST_REQUIRE_EQUAL(str, fmt_str);
+
+    size_t num_elements = expected_strings.size();
+
+    size_t paren_size = expect_parenthesis ? 2 : 0;
+    size_t min_size = paren_size + (x.empty() ? 0 : (num_elements - 1));
+    BOOST_REQUIRE_GE(str.size(), min_size);
+
+    std::string_view sv = str;
+    if (expect_parenthesis) {
+        verify_parenthesis(sv);
+        sv = sv.substr(1, sv.size() - 2);
+    }
+
+    bool first = true;
+    while (!expected_strings.empty()) {
+        sv = trim(sv);
+        if (!std::exchange(first, false)) {
+            sv = cmp_and_remove_prefix(sv, expected_delim);
+        }
+
+        auto s = *expected_strings.begin();
+        sv = cmp_and_remove_prefix(sv, s);
+        expected_strings.erase(expected_strings.begin());
+    }
+
+    BOOST_REQUIRE(sv.empty());
+}
+
+template <std::ranges::range Range>
+void test_format_range(const char* desc, Range x, std::unordered_set<std::string> expected_strings, std::string expected_delim = ",", bool expect_parenthesis = true) {
+    auto str = seastar::format("{}", x);
+    BOOST_TEST_MESSAGE(format("{}: {}", desc, str));
+
+    auto fmt_str = fmt::format("{}", x);
+    BOOST_REQUIRE_EQUAL(str, fmt_str);
+
+    size_t num_elements = expected_strings.size();
+
+    size_t paren_size = expect_parenthesis ? 2 : 0;
+    size_t min_size = paren_size + (x.empty() ? 0 : (num_elements - 1));
+    BOOST_REQUIRE_GE(str.size(), min_size);
+
+    std::string_view sv = str;
+    if (expect_parenthesis) {
+        verify_parenthesis(sv);
+        sv = sv.substr(1, sv.size() - 2);
+    }
+
+    bool first = true;
+    while (!expected_strings.empty()) {
+        sv = trim(sv);
+        if (!std::exchange(first, false)) {
+            sv = cmp_and_remove_prefix(sv, expected_delim);
+        }
+
+        for (auto it = expected_strings.begin(); it != expected_strings.end(); ++it) {
+            if (sv.starts_with(*it)) {
+                sv = cmp_and_remove_prefix(sv, *it);
+                expected_strings.erase(it);
+                break;
+            }
+        }
+    }
+
+    BOOST_REQUIRE(sv.empty());
+}
```

Comment:
Philosophical nitpick (feel free to override):
When you write such long utility functions in a test, it becomes a very real possibility that the test function itself has a bug. I admit the possibility is slim (to fail to catch a bug you would need to create a bug in the original code and exactly the same bug in the same code), but I think it would have been easier to just replace:
```c++
    auto vector = std::vector<int>({1, 2, 3});
    test_format_range("vector", vector, std::vector<std::string>({"1", "2", "3"}));
```
By
```
    auto vector = std::vector<int>({1, 2, 3});
    test_format_range("vector", vector, "1, 2, 3"));
```
With two benefits: the utility function is much shorter, and the reader of the test sees immediately what the printed format actually looks like - are there parentheses or not? Which parentheses?

Reply:
Oh, what I said here only applies to ordered collections. In a set theoretically you can get any order (although in practice the order would be the same every time). So maybe you need a separate test function for unordered collection. Oh, well... So maybe your approach is as good as it gets.